File: HornerAlgo.cs
namespace FloatingPointArithmetic;

public class HornerAlgo
{
    public static (double, double) HornerAlgorithm(double[] a, double x)
    {
        var n = a.Length;
        var r = a[n - 1];
        var s = Math.Abs(a[n - 1]) / 2;
        var u = 1 / 2d * Utils.Ulp(1);
        for (int i = n-1; i > 1; i--)
        {
            r = r * x + a[i-1];
            s = s * Math.Abs(x) + Math.Abs(r);
        }

        r = r * x + a[0];
        var b = 2 * (s * Math.Abs(x) + Math.Abs(r));
        b = u * b / (1 - (2 * (n-1) - 1)*u) ;
        return (r, b);
    }
    
    public static (float, float) HornerAlgorithm(float[] a, float x)
    {
        var n = a.Length;
        var r = a[n - 1];
        var s = Math.Abs(a[n - 1]) / 2;
        var u = 1 / 2f * Utils.Ulp(1f);
        for (int i = n-1; i > 1; i--)
        {
            r = r * x + a[i-1];
            s = s * Math.Abs(x) + Math.Abs(r);
        }

        r = r * x + a[0];
        var b = 2 * (s * Math.Abs(x) + Math.Abs(r));
        b = u * b / (1 - (2 * (n-1) - 1)*u) ;
        return (r, b);
    }
}

File: Exam/HornerAlgo.fs
namespace Exam

open System

module HornerAlgo =

    let prioriSum (a: double[], x: double) =
        let n = a.Length - 1
        let mutable e = abs (a[n])

        for i = n - 1 downto 0 do
            e <- e * abs x + abs (a[i])
        e

    let private calculatePolynomial (a: double[], x: double) =
        let n = a.Length - 1
        let mutable r: double = a[n]
        let mutable s: double = abs(a[n]) / double 2

        for i = n - 1 downto 1 do
            r <- ((r * x) + a[i])
            s <- ((s * abs(x)) + abs(r))

        r <- ((r * x) + a[0])

        let mutable b: double =
            (double 2 * double s * double (abs(x)) + double (abs(r)))

        b <- Ulp.getU * (b / (double 1 - double (2 * n - 1) * Ulp.getU))
        r, b

    let calculate (a: double[], x: double) =
        match a.Length with
        | 0 -> double 0, double 0
        | 1 -> a[0], 0
        | _ -> calculatePolynomial (a, x)
